lib/writable.js

```diff
@@ -1,69 +1,227 @@
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    defaultEncoding = 'utf-8',
+    highWaterMark = DEFAULT_HIGH_WATERMARK
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+
+    if (!this._write && !this._writev) {
+      throw new Error('[WRITE_AND_WRITEV_NOT_IMPLEMENTED]');
+    }
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.highWaterMark = highWaterMark;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(highWaterMark);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (!this.corked && this.writableLength) {
+      this.writing = true;
+      this.doWrite(this.buffer.getBuffer());
+    }
+  }
+
+  setDefaultEncoding(encoding) {
+    this.defaultEncoding = encoding;
+  }
+
+  [errorFuncName](err, cb) {
+    if (typeof cb === 'function') {
+      process.nextTick(cb, err);
+    }
+    this.emit('error', err);
+  }
+
+  pipe() {
+    const err = new Error('[ERR_STREAM_CANNOT_PIPE]');
+    this[errorFuncName](err);
+  }
+
+  validChunk(chunk, cb) {
+    let error;
+    if (chunk === null) {
+      error = new Error('[ERR_STREAM_NULL_VALUES]');
+    } else if (!Buffer.isBuffer(chunk)) {
+      error = new Error('[ERR_INVALID_ARG_TYPE]');
+    }
+
+    if (error) {
+      this[errorFuncName](error, cb);
+      return false;
+    }
+    return true;
+  }
+
+  write(chunk, encoding, cb) {
+    if (typeof encoding === 'function') {
+      cb = encoding;
+      encoding = null;
+    }
+
+    if (this.ended) {
+      const err = new Error('[ERR_STREAM_WRITE_AFTER_END]');
+      this[errorFuncName](err, cb);
+      return;
+    }
+
+    if (this.destroyed) {
+      const err = new Error('[ERR_STREAM_DESTROYED]');
+      this[errorFuncName](err, cb);
+      return;
+    }
+
+    if (!encoding) {
+      encoding = this.defaultEncoding;
+    }
+
+    if (typeof chunk === 'string') chunk = Buffer.from(chunk, encoding);
+
+    if (!this.validChunk(chunk, cb)) return;
+    if (cb) this.callbacks.push(cb);
+
+    if (this.corked || this.writing) {
+      this.buffer.write(chunk, encoding);
+    } else {
+      this.writing = true;
+      this.doWrite(chunk);
+    }
+  }
+
+  doWrite(data) {
+    const callbacks = this.callbacks;
+    this.callbacks = [];
+
+    const cb = () => {
+      callbacks.forEach(cb => cb());
+      if (!this.writableLength) {
+        this.writing = false;
+        if (this.ending) {
+          this.ended = true;
+          this.emit('finish');
+        }
+        return;
+      }
+      this.doWrite(this.buffer.getBuffer());
+    };
+
+    if (this._writev) {
+      if (Array.isArray(data) && data.length > 1) this._writev(data, cb);
+      else this._write(data, cb);
```

Comment:
Does this mean that both _write and _writev are required?
i.e. if there is _writev but we data is not an array then this will cause undefined is not a function if _write is not present.